### PR TITLE
KAFKA-7778: Add KTable.suppress to Scala API

### PR DESCRIPTION
**Pull request by casey-green**
### Detailed description

* Adds `KTable.suppress` to the Scala API.
* Fixed `count` in `KGroupedStream`, `SessionWindowedKStream`, and `TimeWindowedKStream` so that the value serde gets passed down to the `KTable` returned by the internal `mapValues` method.
* `Suppress` API support for Java 1.8 + Scala 2.11

### Testing strategy

I added unit tests covering:
1. Windowed `KTable.count.suppress` w/ `Suppressed.untilTimeLimit`
2. Windowed `KTable.count.suppress` w/ `Suppressed.untilWindowCloses`
3. Non-windowed `KTable.count.suppress` w/ `Suppressed.untilTimeLimit`
4. Session-windowed `KTable.count.suppress` w/ `Suppressed.untilWindowCloses`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala [hunk 1]
```diff
@@ -58,11 +58,8 @@ class TimeWindowedKStream[K, V](val inner: TimeWindowedKStreamJ[K, V]) {
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.TimeWindowedKStream#count`
    */
-  def count()(implicit materialized: Materialized[K, Long, ByteArrayWindowStore]): KTable[Windowed[K], Long] = {
-    val c: KTable[Windowed[K], java.lang.Long] =
-      inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayWindowStore]])
-    c.mapValues[Long](Long2long _)
```
Comment (casey-green) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala]: I for the life of me could not figure out how to pass `materialized` to `mapValues` here. It would either not compile or cause cast exceptions at runtime.

**Comment by casey-green** [comment]
Seems there is a compilation issue w/ `Suppressed.untilTimeLimit`, `Suppressed.untilWindowCloses`, and `BufferConfig.unbounded` in JDK 8 + Scala 2.11.2 related to static methods in Java interfaces (see https://github.com/scala/scala/pull/3989).

```
08:35:04 > Task :streams:streams-scala:compileTestScala
08:35:04 Pruning sources from previous analysis, due to incompatible CompileSetup.
08:35:06 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KTableTest.scala:152: Static methods in interface require -target:jvm-1.8
08:35:06     val suppression = Suppressed.untilTimeLimit[Windowed[String]](Duration.ofSeconds(2L), new StrictBufferConfigImpl())
08:35:06                                                                  ^
08:35:06 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KTableTest.scala:212: Static methods in interface require -target:jvm-1.8
08:35:06     val suppression = Suppressed.untilWindowCloses(new StrictBufferConfigImpl())
08:35:06                                                   ^
08:35:06 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/KTableTest.scala:271: Static methods in interface require -target:jvm-1.8
08:35:06     val suppression = Suppressed.untilTimeLimit[String](Duration.ofSeconds(2L), new StrictBufferConfigImpl())
08:35:06                                                        ^
08:35:06 three errors found
08:35:07 
```

I think there are a few options:
1. Bump support for scala 2.11 from 2.11.2 => 2.11.3.
2. Move `Suppressed.untilTimeLimit` and `Suppressed.untilWindowCloses` out of the `Suppressed` Java interface.
3. Modify my unit tests to `new` the underlying classes created in `Suppressed.until*`. This seems hacky because the classes are inside of `org.apache.kafka.streams.kstream.internals` package (key word `internals`). This will also be confusing for people on scala 2.11.2.

**Comment by bbejeck** [comment]
ping @vvcephei @ableegoldman for reviews

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala [hunk 2]
```diff
@@ -45,11 +45,8 @@ class KGroupedStream[K, V](val inner: KGroupedStreamJ[K, V]) {
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.KGroupedStream#count`
    */
-  def count()(implicit materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = {
-    val c: KTable[K, java.lang.Long] =
-      inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayKeyValueStore]])
-    c.mapValues[Long](Long2long _)
-  }
+  def count()(implicit materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] =
+    inner.aggregate((() => 0L).asInitializer, ((_: K, _: V, agg: Long) => agg + 1L).asAggregator, materialized)
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Why this change? Is that intentional?

Comment (guozhangwang) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Ditto below.

**Review by guozhangwang — COMMENTED** [review]
The suppression API seems right to me but I'm not sure about the other, seemly orthogonal change.

cc @vvcephei for another look.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala (hunk 2 above)
Comment (casey-green) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: It is. Without this change, my unit my unit tests fail with this exception:

```
Caused by: java.lang.ClassCastException: java.lang.Long cannot be cast to [B
	at org.apache.kafka.common.serialization.ByteArraySerializer.serialize(ByteArraySerializer.java:19)
	at org.apache.kafka.streams.kstream.internals.FullChangeSerde$1.serialize(FullChangeSerde.java:73)
	at org.apache.kafka.streams.kstream.internals.FullChangeSerde$1.serialize(FullChangeSerde.java:60)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.buffer(KTableSuppressProcessor.java:96)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.process(KTableSuppressProcessor.java:87)
	at org.apache.kafka.streams.kstream.internals.suppress.KTableSuppressProcessor.process(KTableSuppressProcessor.java:40)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:117)
```

It looks like this is because the value serde is not passed down here: 
* https://github.com/apache/kafka/blob/4824dc994d7fc56b7540b643a78aadb4bdd0f14d/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedTable.scala#L50
* https://github.com/apache/kafka/blob/4824dc994d7fc56b7540b643a78aadb4bdd0f14d/streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala#L51

The `KTable` returned by `mapValues` has a null value serde which causes it to fallback to the configured default (byte array serde in this case), which is incorrect.

If you have other ideas for how to fix this (I tried a few different things) or prefer that I put it in a different PR (since it is sort of orthogonal as you said), let me know!

Comment (guozhangwang) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: cc @vvcephei could you take a look?

Comment (vvcephei) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: That's right @casey-green . As I see it, the two options are:
1. implement a scala-native count (which you've done)
2. propagate the serdes (see below)

Propagating the serdes would look like this:
```scala
  def count()(implicit materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = {
    val javaCountTable: streams.kstream.KTable[K, lang.Long] =
      inner.count(materialized.asInstanceOf[Materialized[K, lang.Long, ByteArrayKeyValueStore]])
    val tableImpl = javaCountTable.asInstanceOf[KTableImpl[K, ByteArrayKeyValueStore, lang.Long]]
    javaCountTable.mapValues[Long](
      (l => long2Long(l)) : ValueMapper[lang.Long, Long],
      Materialized.`with`[K, Long, ByteArrayKeyValueStore](tableImpl.keySerde(), Serdes.Long)
    )
  }
```
... pretty nasty.

To be honest, I was never a fan of the extra `mapValues` node there. It seemed wasteful to me. I think in a vacuum, I prefer your solution.

However, dropping a node from the topology results in compatibility problems. If we were to elide the `mapValues` now, then people would have to perform an application reset when upgrading to this version. We'd probably better avoid that.

Thoughts?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala [hunk 3]
```diff
@@ -58,11 +58,8 @@ class TimeWindowedKStream[K, V](val inner: TimeWindowedKStreamJ[K, V]) {
...
-  }
+  def count()(implicit materialized: Materialized[K, Long, ByteArrayWindowStore]): KTable[Windowed[K], Long] =
```
Comment (vvcephei) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala]: Also worth noting that the same problem afflicts `SessionWindowedKStream.scala`. Can you add coverage for that class in your test as well?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala (hunk 2 above)
Comment (casey-green) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Ah, good point about compatibility! In this case, I think the somewhat nastier code is preferable.

I went ahead and made these changes in each of the three `count` methods.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala (hunk 3 above)
Comment (casey-green) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala]: Good catch. Updated, and added a unit test that is effectively a copy of

https://github.com/apache/kafka/blob/04e206154ac614b7d4d34a7a1b6ba2c882f607b9/streams/src/test/java/org/apache/kafka/streams/kstream/internals/SuppressScenarioTest.java#L463

**Review by vvcephei — APPROVED** [review]
LGTM! Thanks, @casey-green .

**Comment by vvcephei** [comment]
The test results have been lost by now...

Retest this, please.

**Comment by guozhangwang** [comment]
Jenkins failure: `Execution failed for task ':streams:streams-scala:compileTestScala'.`

@casey-green could check if this is related?

**Comment by casey-green** [comment]
@guozhangwang yes, compilation issue was related to my changes. I updated my tests to be compatible w/ Scala 2.11, so both scala 2.12 and 2.11 _should_ work (at least the tests compiled and passed locally 😄).

@vvcephei FYI - if you're using scala 2.11 you can't use the static methods inside the `Suppressed` and `BufferConfig` interfaces because of compatibility issue I describe here: https://github.com/apache/kafka/pull/6314#issuecomment-467058765. The workaround is to use the internal class returned by these static methods, so NBD.

**Comment by vvcephei** [comment]
Hey @casey-green , That's really unfortunate.

So, to clarify, anyone who is using 2.11.3+ can reference the static interface methods, but people using 2.11.2- cannot? We can request to bump the build version up to 2.11.3, but I'd still be concerned that other folks out there would expect our 2.11 artifact to be compatible with all 2.11 versions.

Telling people to reference the internal classes would be a bad precedent, and it may allow them to circumvent the invariants we've established in the builder.

Another alternative would be to create a Scala `Suppressed` builder class for scala developers to use. Then, we can be sure that suppression is clean and easy from both Java and Scala, regardless of version. We could put a "todo" on it to deprecate and remove it once we drop support for 2.11.

What do you think?

**Comment by casey-green** [comment]
> So, to clarify, anyone who is using 2.11.3+ can reference the static interface methods, but people using 2.11.2- cannot?

Actually, I think there's an issue with _any_ minor version of Scala 2.11 in combination with JDK 1.8. (I tried Scala 2.11.3, 2.11.4 etc and none compiled.)

> Telling people to reference the internal classes would be a bad precedent

Completely agree. It felt a little gross using them in my unit tests.

> Another alternative would be to create a Scala `Suppressed` builder class for scala developers to use. 

This seems like the best option to me. Would you prefer I make that change in this PR or create a new one after this one is merged? Happy to do either!

**Comment by vvcephei** [comment]
Ah, thanks for that clarification... I think we'd better do it in this PR, so that the whole "make it work with Scala" feature is in one chunk. The PR's not so big right now that it would justify splitting it up, either.

Off the cuff, I think that we'd basically need a Scala version of everything that's in the Suppressed interface, because we'll also need the equivalent of the static factories for the buffer configurations.

Of course, we need to make sure that everything gets wired up properly, so we'll also need a unit test verifying that it creates the same internal objects as the java builders.

**Comment by casey-green** [comment]
> I think we'd better do it in this PR, so that the whole "make it work with Scala" feature is in one chunk.

Makes sense 👍 

@vvcephei I took a slightly different approach, which you may have an opinion about because I modified some of the java code _slightly_:
1. I removed the type parameter from `BufferConfig`. This type parameter was causing lots of problems on the Scala side, and the semantics appear to be the same without them. It is entirely possible that I'm missing something (future use case, established design pattern, etc), so let me know if this approach would cause issues or is not desirable.
2. Replicated the corresponding static factory methods in scala.

Let me know if you have feedback on this approach, happy to iterate on it.

**Comment by vvcephei** [comment]
Hi @casey-green ,

Thanks for the quick update! I can see where you're coming from, but those type parameters are actually necessary.

The idea is that we want to provide a compile-time guard against using `untilWindowClose` with a non-strict buffer. Allowing a non-strict buffer in that configuration would violate the semantics of the operation (namely, that suppression will definitely not emit until after the window closes).

I hope you don't mind, but I hesitated to just push back without trying to help, as I know it can be very difficult to reconcile java's and scala's type systems under certain circumstances. So, I prepared a mini PR for your branch (https://github.com/casey-green/kafka/pull/1) just as a succinct way to express what I'm thinking (and make sure that it actually compiles, of course).

The recursive type parameter is there so that the chained builder methods can preserve the type of the object (i.e., `withXX` can return the right subclass, as opposed to the interface itself). In Suppressed, the only assertion we really care about making is that `untilWindowCloses` must take a `StrictBufferConfig`. The rest of the `BufferConfigJ` references in Suppressed are free to use a wildcard type parameter, indicating that they don't care what kind of buffer they're getting (when it's a parameter), or that they make no guarantees about the strictness of the buffer that they return (when it's a return type).

The compile-time restriction is nice as a user, but it's a bummer as a library, since it's impossible to write tests that verify the restriction. I.e., we can only test that permitted code compiles, we can't test that non-permitted code doesn't compile (because the test wouldn't compile).

Thanks again for your help. What do you think?

**Comment by casey-green** [comment]
@vvcephei thank you for the quick feedback! I appreciate the responsiveness. Sorry this PR has probably taken up a lot of your time at this point, so thank you for bearing with me.

>  it can be very difficult to reconcile java's and scala's type systems under certain circumstances

Agreed, and generics/type bounds are especially a PITA to reconcile between java/scala.

> The idea is that we want to provide a compile-time guard against using untilWindowClose with a non-strict buffer. Allowing a non-strict buffer in that configuration would violate the semantics of the operation...
> The recursive type parameter is there so that the chained builder methods can preserve the type of the object...

Thanks for the explanation, that makes sense. Let me take your suggestions and iterate on this; I'll move any questions/discussion to the PR you made against my branch.

**Comment by vvcephei** [comment]
Hey @casey-green ,

No worries about the back-and-forth! I'm happy to help get this PR merged.

FYI, @bbejeck independently noticed the same awkward extra processor node in `count`: https://issues.apache.org/jira/browse/KAFKA-8101

**Comment by casey-green** [comment]
@vvcephei ready for another review. Implemented the `EagerBufferConfig` marker interface approach as discussed in https://github.com/casey-green/kafka/pull/1.

**Review by vvcephei — APPROVED** [review]
Looks good to me! Thanks @casey-green 

Perhaps @guozhangwang or @mjsax can take another look.

**Comment by casey-green** [comment]
Thanks for the reviews @vvcephei! Any more thoughts here from @guozhangwang or @mjsax ?

**Comment by guozhangwang** [comment]
@vvcephei @casey-green  if we are introducing a `EagerBufferConfig` then we should create a KIP since it is public interface ..

**Comment by casey-green** [comment]
@guozhangwang @vvcephei could we update the existing KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables rather than creating a new one?

**Comment by guozhangwang** [comment]
Maybe we can, since KIP-328 is not completely done yet (regarding the disk-spilling) and hence it would be spanning over two releases anyways. If this can be done before the coming 2.3 release feature deadline I think we can piggy-back this change as well. cc @vvcephei

**Comment by casey-green** [comment]
Can one of you grant me access to that wiki page, and I'll make the update? wiki username = `casey-green`

**Comment by guozhangwang** [comment]
@casey-green I've added you to the wiki space, should be able to modify the KIP-328 wiki now. 

After all the changes being made, and after the spill-to-disk PR is finalized, I think @vvcephei can send out an email on the mailing thread to summarize all the tweaks we have been since the vote is completed.

**Comment by casey-green** [comment]
Thanks @guozhangwang! KIP updated: https://cwiki.apache.org/confluence/pages/diffpagesbyversion.action?pageId=87295409&selectedPageVersions=25&selectedPageVersions=24

Let me know if there is anything else to do on my end.

**Comment by guozhangwang** [comment]
Made another pass. LGTM! Thanks for the contribution @casey-green .

**Comment by guozhangwang** [comment]
@vvcephei could you send an notification on the KIP thread about the augment, and also file a PR for upgrade_guide docs as well?

**Comment by jurgispods** [comment]
I'm using the latest 2.2.1 Maven release of `kafka-streams-scala`, but I'm getting a

```
value suppress is not a member of org.apache.kafka.streams.scala.kstream.KTable[org.apache.kafka.streams.kstream.Windowed[Key], Value]
```

Has this change not yet made it upstream or am I missing something else?

**Comment by casey-green** [comment]
@pederpansen this will be in the 2.3.0 release https://github.com/apache/kafka/pull/6586
